include/ur_client_library/ur/watchdog.h

```diff
@@ -0,0 +1,92 @@
+// -- BEGIN LICENSE BLOCK ----------------------------------------------
+// Copyright 2023 Universal Robots A/S
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//    * Redistributions of source code must retain the above copyright
+//      notice, this list of conditions and the following disclaimer.
+//
+//    * Redistributions in binary form must reproduce the above copyright
+//      notice, this list of conditions and the following disclaimer in the
+//      documentation and/or other materials provided with the distribution.
+//
+//    * Neither the name of the {copyright_holder} nor the names of its
+//      contributors may be used to endorse or promote products derived from
+//      this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+// -- END LICENSE BLOCK ------------------------------------------------
+
+#pragma once
+
+#include "ur_client_library/comm/control_mode.h"
+#include <chrono>
+
+namespace urcl
+{
+/*!
+ * \brief Watchdog class containing a timeout configuration
+ *
+ * This watchdog is used to configure the read timeout for the reverse socket running in the external control script.
+ * The read timeout is the number of milliseconds until the read action times out. A timeout of 0 or negative number
+ * indicates that the function should not return until a read is completed, this will make the read function blocking on
+ * the robot. This can be set using the function off().
+ *
+ */
+class Watchdog
```

Comment:
While writing my previous comments about `std::chrono` I think I didn't realize what really bugged be: The `Watchdog` class isn't really a watchdog. When I read "watchdog" I expect an object that has a timeout, actually "watches" some process and raises something if that timeout is exceeded.

However, this class is merely a timeout itself with a couple of fancy construction functions. We could have the timeout as a single `std::chrono` object and pass that down to the write functions. The constructor methods could be simple macros e.g. `#define READ_TIMEOUT_OFF std::chrono::milliseconds(0)`. I'm not saying, we should, but we could without changing much.

The `verifyWatchdogTimeout()` method wouldn't loose much from its benefit if it was just a standalone helper function that takes the control_mode, step_time and timeout as arguments. Since we are calling this in each write method of the reverse interface manually, this wouldn't actually be much worse than the current approach.

I'm not sure whether I would vote for removing the class, but if we keep it I would like to propose a more meaningful name for the class. For example:
- `ControlTimeout`
- `ControlReadTimeout`
- `ControlTimeoutValidator`
- `ControlReadTimeoutValidator`
- `TimeoutValidator`

I think my favorite would be `ControlReadTimeout` since that best describes the use case.

Reply:
The reason why I have suggested this class/type is to avoid people use zero as an argument for the time and expected that this is the smallest timeout value where it in reality completely turn off and make the socket command a blocking call.
While still support that it could be a blocking call waiting for a response when the robot e.g. are standing still.

I agree regarding the naming, maybe just call it Timeout which is what it is.

Reply:
Agreed on `RobotReceiveTimeout`